Content.Client/Overlays/ComponentAddedOverlaySystemBase.cs

```diff
@@ -0,0 +1,115 @@
+
+            SubscribeLocalEvent<T, ComponentInit>(OnInit);
+            SubscribeLocalEvent<T, ComponentRemove>(OnRemove);
+
+            SubscribeLocalEvent<PlayerAttachedEvent>(OnPlayerAttached);
+            SubscribeLocalEvent<PlayerDetachedEvent>(OnPlayerDetached);
+
+            SubscribeLocalEvent<T, GotEquippedEvent>(OnCompEquip);
+            SubscribeLocalEvent<T, GotUnequippedEvent>(OnCompUnequip);
+
+            SubscribeLocalEvent<RoundRestartCleanupEvent>(OnRoundRestart);
+
+            _invSystem = _entityManager.System<InventorySystem>();
+        }
+
+        public void ApplyOverlay(T component)
+        {
+            IsActive = true;
+            OnApplyOverlay(component);
+        }
+
+        public void RemoveOverlay()
+        {
+            IsActive = false;
+            OnRemoveOverlay();
+        }
+
+        protected virtual void OnApplyOverlay(T component) { }
+
+        protected virtual void OnRemoveOverlay() { }
+
+        private void OnInit(EntityUid uid, T component, ComponentInit args)
+        {
+            if (_player.LocalPlayer?.ControlledEntity == uid)
+            {
+                ApplyOverlay(component);
+            }
+        }
+
+        private void OnRemove(EntityUid uid, T component, ComponentRemove args)
+        {
+            if (_player.LocalPlayer?.ControlledEntity == uid)
+            {
+                RemoveOverlay();
+            }
+        }
+
+        private void OnPlayerAttached(PlayerAttachedEvent args)
+        {
+            if (TryComp<T>(args.Entity, out var component))
+                ApplyOverlay(component);
+
+            if (TryComp(args.Entity, out InventoryComponent? inventoryComponent)
+                && _invSystem.TryGetSlots(args.Entity, out var slotDefinitions, inventoryComponent))
+            {
+                foreach (var slot in slotDefinitions)
+                {
+                    if (_invSystem.TryGetSlotEntity(args.Entity, slot.Name, out var itemUid)
+                        && TryComp(itemUid.Value, out component))
+                    {
+                        OnCompEquip(itemUid.Value, component, new GotEquippedEvent(args.Entity, itemUid.Value, slot));
+                    }
+                }
+            }
+        }
+
+        private void OnPlayerDetached(PlayerDetachedEvent args)
+        {
+            RemoveOverlay();
+        }
+
+        private void OnCompEquip(EntityUid uid, T component, GotEquippedEvent args)
+        {
+            if (!TryComp<ClothingComponent>(uid, out var clothing)) return;
+
+            if (args.Equipee != _player.LocalPlayer?.ControlledEntity) return;
+
+            if (!clothing.Slots.HasFlag(args.SlotFlags)) return;
+
+            ApplyOverlay(component);
+        }
+
+        private void OnCompUnequip(EntityUid uid, T component, GotUnequippedEvent args)
+        {
+            if (args.Equipee != _player.LocalPlayer?.ControlledEntity) return;
+
+            RemoveOverlay();
+        }
```

Comment:
This will still cause issues if an entity that innately has an overlay equips and unequips something that would usually grant that overlay.

This should probably be changed into a proper overlay granting system that does the following:
- When overlay-granting items are (un)equipped, or if relevant components are added/removed from any entity, then refresh the available overlay(s).
- Also refresh whenever the currently controlled entity changes.
- Overlays are refreshed by raising a get-overlay event directed at the player's current entity. The event can just get relayed via the normal equipment relay stuff.

That should work for anything that uses simple overlays, but in this case you also need each component to be able to modify the overlay because of the `List<string> DamageContainers`. I.e., if a player has both a hardsuit helmet and some HUD glasses that have different container IDs, both should get shown.

I guess that could be fixed by making the get-overlay event return a `Dictionary<Type, Overlay>`, and subscribers that want to add a duplicate overlay can handle it however they see fit. So for the previous example, the second subscriber could just add additional IDs to the overlay that is already getting returned by the event?